### DISCUSSION TITLE
How can I use Redis inside getStaticProps if I have to import the function that will use it in a file that renders on the front-end?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#18230] [Help] [answered]
```js
//index.js
import { functionThatUsesRedis } frrom '../functionThatUsesRedis.js';
...
...
export const getStaticProps = async(context) => {
  const data = await functionThatUsesRedis();
}
```
```js
//functionThatUsesRedis.js
const Redis = require('ioredis');
const client = new Redis;
```
I want to be able to use `functionThatUsesRedis` so that I can populate my static props. `getStaticProps` only runs in the server, but the **index.js file gets dragged into the browser and since I declated the import of `functionThatUsesRedis` at the top of it, that too will be put in there.** This all results in the following error:
```
error - ./node_modules/ioredis/built/cluster/ClusterOptions.js:3:0
Module not found: Can't resolve 'dns'
```
Naturally. It's trying to load Redis inside the client. Still, is there on way to lazy-load or tell Next that I need `functionThatUsesRedis ` but only when it's time to run `getStaticProps`?

**Comment by jamesmosier** [comment]
> the index.js file gets dragged into the browser

Could you clarify this? 

---

`getStaticProps` removes any imports on the client side that you use inside `getStaticProps`, getServerSideProps`, & `getStaticPaths`. You can see this in action using [this tool](https://next-code-elimination.now.sh/).

**Reply by ghost** [reply]
But that's the thing. `index.js` is a page on the client-side, however, its `getStaticProps` only runs server side, but as I said, I'm importing `functionThatUsesRedis` inside of `index.js` and that uses Redis. Redis isn't available on the client-side however so it errors out.

What I would need here is to `import { functionThatUsesRedis } frrom '../functionThatUsesRedis.js';` **inside `getStaticProps`.**

**Reply by ljosberinn** [reply]
No, that is precisely _not_ the case, as the [docs](https://nextjs.org/docs/basic-features/data-fetching#write-server-side-code-directly) mention. See [here](https://next-code-elimination.now.sh/) for a comparison of which code will be actually bundled.

If you have some serverside code ending up in the bundle, your imports are either impure or you export both client side only and serverside only code that's used outside of get...Props too.

**Reply by ghost** [reply]
But how could this be? The tool works - it's not adding the `functionThatUsesRedis` to the client bundle, yet my Next.js does include id (canary).

**Reply by ghost** [reply]
It's also weird, because it only does this with that specific function only, but I have another function which depends on `process.env`:

```js
/**
 * Retrieves data about featured videos.
 */
export const getFeaturedVideos = async() => {
    try {
        const response = await fetch(
            `${process.env.NODE_SERVER}/api/featuredVideos`,
            {
                method: 'get',
            }
        );

        let data = response.json();

        if (response.status >= 400 && response.status < 600) {
            throw Error(response.statusText);
        }

        return data;
    } catch(error) {
        throw Error(`Error getting featured videos: ${error.message}`);
    }
}
```
And this also clearly only works on the server-side and it only runs on the server, yet the other one with Redis errors out.

**Reply by ljosberinn** [reply]
I'm fairly confident that your function must be included erroneously within some client-side only code. There _may_ also be an issue if it is within a module that also exports something you definitely only use client-side; I've run into a similar issue setting up Sentry.
Especially since you mention it's an issue with _that_ function makes me conclude it's that situation.

**Reply by ghost** [reply]
This was indeed an issue with my linter. It just didn't catch (for some weird reason) that function that I imported as not used and it wasn't greyed out, nor did the error give any signal as to where the issue is precisely.

So, in short: when you're importing a function, if you don't use it anywhere, JS sees it as "global scope" of some sorts, however, if it's used inside `getStaticProps`, it inherits the (server-side) running context and isn't called on the client-side.

**Comment by ghost** [comment, marked as answer]
This was indeed an issue with my linter and VSCode, sorry for the trouble. It just didn't catch (for some weird reason) that function that I imported as not used and it wasn't greyed out, nor did the error give any signal as to where the issue is precisely.

So, in short: when you're importing a function, if you don't use it anywhere, JS sees it as "global scope" of some sorts, however, if it's used inside getStaticProps, it inherits the (server-side) running context and isn't called on the client-side.
